### ISSUE TITLE
WHEELS: provide older version of dev wheels rather than overwriting the latest

### ISSUE THREAD

**Issue by bsipocz** [numpy/numpy#24424]
Once a downstream CI starts to fail due to a new nightly version, it would be extremely helpful to be able to go back a few versions for debugging and testing.

Somewhat related (but I'm happy to open a separate issue for it), is  that while I understand that currently there are technical limitations, having the git hash in the version would be super useful, too. But even having the date would be nicer than the extremely generic `dev0`

**Comment by mattip** [comment]
FWIW, SciPy uses the same versioning (or non-versioning) of wheels uploaded to https://anaconda.org/scientific-python-nightly-wheels

**Comment by bsipocz** [comment]
I suppose it's the same underlying problem with meson, so once there is a solution for numpy it can be ported to scipy, too. Or do you want me to open this same issue at scipy as well?

**Comment by charris** [comment]
I could be wrong, but my understanding is that labeling the wheels is a problem, but the numpy version should be more informative. That would be new, I think. Looking in one of yesterday's wheels I see
```
version = "2.0.0.dev0+git20230814.2c45a56"
```

which is what `numpy.__version__` should show. I haven't tried it, though.

**Comment by bsipocz** [comment]
Yes, I wasn't clear I suppose. I see that version in CI where the script explicitly calls `numpy.__version__`, but e.g. `pip freeze` still shows only `2.0.0.dev0`, neither it is shown in the filename.

**Comment by mattip** [comment]
There is a long discussion at mesonbuild/meson-python#159 about allowing the pyproject.toml `version` attribute to be dynamic. If I understand the last few comments correctly, this feature is still not available in meson-python. 

> Or do you want me to open this same issue at scipy as well?

No, no need for a new issue there. I was hoping they had solved the problem for us

**Comment by mattip** [comment]
>  otherwise they accumulate rapidly 

It turns out the site has a [retention policy](https://github.com/scientific-python/specs#261) enforced by a [github action](https://github.com/scientific-python/upload-nightly-action/blob/main/.github/workflows/remove-wheels.yml): every day there is a cron job that leaves the last 5 wheel versions and removes artifacts that are more than 30 days old.

**Comment by bsipocz** [comment]
Oh, and building from source: it's a me problem, and not that relevant here (well, someone else has the same problem): I still can't build scipy from source, meson has an issue with my pyenv pythons and paths, etc. and I haven't managed to build it from source for a few months now. Numpy is similarly painful, but we found one workaround for it at Scipy23, that kind of works (in practice it's not at all convenient).

So, at the end, I switched to using the nighlies, too when I need something "close enough to dev" 🤷‍♀️

**Comment by ngoldbaum** [comment]
> meson has an issue with my pyenv pythons and paths

Not sure about SciPy, but I've been building NumPy with a pyenv python since the meson transition with only a few bumps at first and no issues for a while now. I'm on Linux though, if that matters. If you think looking at my setup would be helpful feel free to ping me.

**Comment by bsipocz** [comment]
Thanks, I'll take you up on it and follow up. (The current workaround is to work with a temporary venv, so it is really not convenient, and even the non meson <3.12 ways are not working)

**Comment by rgommers** [comment]
I would like to keep the metadata in `pyproject.toml` fully static, as discussed before. If it's useful to append a build number for nightlies (so the wheel name contains `2.0.0.dev0-N` with `N` an integer like the number of commits or some such thing), this could be done post-hoc in the CI script for uploading.

> Thanks, I'll take you up on it and follow up.

Thanks, it'd be great to sort this out. I wasn't aware of any problem or bug report. This should work, happy to help too if needed.
